src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java

```diff
@@ -1147,4 +1148,229 @@ public void testCleanerRecoverListCount() throws Exception {
+            .build();
+    offset = 0;
+    // Create sink record with same offset
+    SinkRecord record2 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+    service2.insert(record2);
+    // The record should not be inserted in the buffer
+    Assert.assertTrue(
+        ((SnowflakeSinkServiceV1) service2)
+            .isPartitionBufferEmpty(SnowflakeSinkServiceV1.getNameIndex(topic, partition)));
+  }
+
+  @Ignore
+  @Test(expected = SnowflakeKafkaConnectorException.class)
+  // files write to FDB for an older clientSequencer should be discarded
+  public void testFileDiscardWithExactlyOnceSemantics() throws Exception {
+    conn.createTable(table);
+    conn.createStage(stage);
+    SnowflakeSinkService service =
+        SnowflakeSinkServiceFactory.builder(conn)
+            .setRecordNumber(1)
+            .setDeliveryGuarantee(
+                SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.EXACTLY_ONCE)
+            .addTask(table, topic, partition)
+            .build();
+
+    SnowflakeConverter converter = new SnowflakeJsonConverter();
+    SchemaAndValue input =
+        converter.toConnectData(topic, "{\"name\":\"test\"}".getBytes(StandardCharsets.UTF_8));
+
+    SnowflakeIngestionService ingestionService = conn.buildIngestService(stage, pipe);
+
+    long offset = 0;
+    // Create sink record
+    SinkRecord record1 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+
+    service.insert(record1);
+    // Confirm the record is flushed to stage
+    TestUtils.assertWithRetry(
+        () ->
+            conn.listStage(
+                        stage,
+                        FileNameUtils.filePrefix(TestUtils.TEST_CONNECTOR_NAME, table, partition))
+                    .size()
+                == 1,
+        5,
+        4);
+
+    // Make sure correct client sequencer
+    ClientStatusResponse clientStatusResponse = ingestionService.getClientStatus();
+    assert clientStatusResponse.getClientSequencer() == 0;
+    assert clientStatusResponse.getOffsetToken() == null;
+
+    // initialize a new sink service without closing the old service yet
+    SnowflakeSinkService service2 =
+        SnowflakeSinkServiceFactory.builder(conn)
+            .setRecordNumber(1)
+            .setDeliveryGuarantee(
+                SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.EXACTLY_ONCE)
+            .addTask(table, topic, partition)
+            .build();
+    offset = 1;
+    // Create sink record with same offset
+    SinkRecord record2 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+    // Should update the client sequencer
+    service2.insert(record2);
+    // call snowpipe ingest api
+    service.callAllGetOffset();
+
+    service.closeAll();
+    service2.closeAll();
```

Comment:
but we are changing the offset here to 1? So it should ingest it?

Reply:
Sorry, the comment wasn't very clear, the old service calls get offset, so the api call should throw exception due to outdated client sequencer.